core/resourceid/module_id.go

```diff
@@ -0,0 +1,47 @@
+package resourceid
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// EncodeModule JSON marshals and base64-encodes a module ID from its typename
+// and data.
+func EncodeModule(typeName string, value any) (string, error) {
+	jsonBytes, err := json.Marshal(value)
+	if err != nil {
+		return "", fmt.Errorf("failed to json marshal %T: %w", value, err)
+	}
+	idEnc := base64.StdEncoding.EncodeToString(jsonBytes)
+	return fmt.Sprintf("moddata:%s:%s", typeName, idEnc), nil
+}
+
+// DecodeModule base64-decodes and JSON unmarshals an ID, returning the module
+// typename and its data.
+func DecodeModule(rest string) (string, any, error) {
```

Comment:
nit: These names feel a little confusing, since they sound like they encode/decode the module itself, but they actually encode/decode a typed object from a module.

Maybe something like this?

```suggestion
func DecodeModuleObject(rest string) (string, any, error) {
```

(and `EncodeModuleObject` in turn)